overlord/devicestate/devicestate.go

```diff
@@ -1170,13 +1178,29 @@ func checkForRequiredSnapsNotInModel(model *asserts.Model, vSets *snapasserts.Va
 	return nil
 }
 
+func checkAllSnapsFromModelInValidationSets(model *asserts.Model, vSets *snapasserts.ValidationSets) error {
+	if len(vSets.Keys()) == 0 {
+		return nil
+	}
+
+	for _, sn := range model.AllSnaps() {
+		if !vSets.SnapConstrained(sn) {
+			return fmt.Errorf("snap presence is not constained by a validation set: %s", sn.SnapName())
+		}
```

Comment:
did we discuss this? it seems over-constraining but maybe I'm missing something. It's especially strange with all local snaps

Reply:
From the spec: "Snapd will check that all snaps in the model (including mode-specific ones) are part of the validation sets and error otherwise." I took this literally as they must be in the validation set. Maybe the intention was that the snaps must follow the rules in the validation set.

Reply:
I've removed that restriction.